### PR TITLE
Add a function that generates an SP initiated SSO request

### PR DESCRIPTION
**Pull request by fumieval**
This PR adds a basic support for SP-initiated flow. I confirmed that it works with Google Workspace

### CODE REVIEW

File: wai-saml2.cabal [hunk 1]
```diff
@@ -20,6 +20,7 @@ build-type:     Simple
 extra-source-files:
     README.md
     CHANGELOG.md
+    tests/data/google.xml.base64
```
Comment (mbg) [inline wai-saml2.cabal]: This file doesn't seem to be included in the PR - did you mean to include it or exclude this line?

File: src/Network/Wai/SAML2/Request.hs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import qualified Data.Map.Strict as Map
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+
+-- | Parameters for SP-initiated SSO
+data Request = Request
+  { requestIssuer :: T.Text
+  , requestAllowCreate :: Bool
+  , requestNameIDFormat :: T.Text
+  }
```
Comment (mbg) [inline src/Network/Wai/SAML2/Request.hs]: `Request` is a pretty generic name. I am wondering if it might be better to rename this to e.g. `AuthnRequest`, which would also better correspond to the XML element name?

There is also no documentation for the fields here. I won't hold up the PR because of this, but it would be nice. Again, see the [contribution guidelines](https://github.com/mbg/wai-saml2/blob/master/CONTRIBUTING.md#documentation) with respect to documentation. In particular, it might also be nice to extend the README with documentation on how to use `issueRequest`, or provide module-level documentation.

File: src/Network/Wai/SAML2/Request.hs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+-- | Generated a base64-encoded AuthnRequest for SP initiated SSO.
+-- See also: http://docs.oasis-open.org/security/saml/Post2.0/sstc-saml-tech-overview-2.0-cd-02.html#5.1.2.SP-Initiated%20SSO:%20%20Redirect/POST%20Bindings|outline
+issueRequest :: Request -> IO B.ByteString
+issueRequest Request{..} = do
+    now <- getCurrentTime
+    let timestamp = T.pack $ formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%SZ" now
```
Comment (mbg) [inline src/Network/Wai/SAML2/Request.hs]: Could this be using the format from https://github.com/mbg/wai-saml2/blob/d4191b5d2d22e36746db67517b60cccf6d40f7a0/src/Network/Wai/SAML2/XML.hs#L66 to avoid the string constant here?

File: src/Network/Wai/SAML2/Request.hs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    ident <- T.decodeUtf8 . Base64.encode <$> getRandomBytes 64
```
Comment (mbg) [inline src/Network/Wai/SAML2/Request.hs]: According to [the SAML2 specification](https://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf) (Section 3.2.1, Line 1467 onward), we must check that the identifier in the response matches the identifier in the request. I don't see a way to do that with the current implementation, since `ident` isn't returned to the caller in an easily accessible way (or provided by the caller).

**Review by mbg — COMMENTED** [review]
It looks like most of new the Haskell file is formatted using Ormolu (or similar), although there is some inconsistent indentation. It would be good if the formatting could be more in line with the other files, as per the [contribution guidelines](https://github.com/mbg/wai-saml2/blob/master/CONTRIBUTING.md#style) for style.

File: wai-saml2.cabal (hunk 1 above)
Comment (fumieval) [reply wai-saml2.cabal]: Hmm, it looks like `hpack` added this line, which isn't intended

File: src/Network/Wai/SAML2/Request.hs (hunk 4 above)
Comment (fumieval) [reply src/Network/Wai/SAML2/Request.hs]: That's a good point. I split the function into two; `issueAuthnRequest :: Text -> IO AuthnRequest` generates a default `AuthnRequest` along with a timestamp and ID, and `renderAuthnRequest :: AuthnRequest -> ByteString`. This also makes the datatype correspond better with AuthnRequest on specification.

File: src/Network/Wai/SAML2/Request.hs (hunk 2 above)
Comment (fumieval) [reply src/Network/Wai/SAML2/Request.hs]: I renamed the datatype and added comments to each field

File: src/Network/Wai/SAML2/Request.hs (hunk 3 above)
Comment (fumieval) [reply src/Network/Wai/SAML2/Request.hs]: fixed

Comment (mbg) [reply src/Network/Wai/SAML2/Request.hs]: It seems that you still re-define the format in `Network.Wai.SAML2.Request`. What I meant was that you should use `timeFormat` from `Network.Wai.SAML2.XML` to avoid duplication. Sorry if that wasn't clear. I appreciate that it's not currently exported from `Network.Wai.SAML2.XML`, so the options are:

- Adding it to `Network.Wai.SAML2.XML`'s export list.
- Defining a `formatUTCTime` function in `Network.Wai.SAML2.XML` and exporting that.

When I was looking over this, I was also wondering if we could just use [the ISO8601 functions from the `time` package](https://hackage.haskell.org/package/time-1.12.2/docs/Data-Time-Format-ISO8601.html) to begin with, but I didn't find the SAML2 specification to be super clear whether it fully conforms to ISO8601.

File: src/Network/Wai/SAML2/Request.hs (hunk 4 above)
Comment (mbg) [reply src/Network/Wai/SAML2/Request.hs]: Thanks, that seems like a better way to organise this.

**Comment by mbg** [comment]
@fumieval Thanks for your work on this and the changes you have made in response to the review. It seems that the formatting and documentation are still outstanding. Is that something you are still planning to do? If not, I can deal with the documentation and re-format the code.

File: src/Network/Wai/SAML2/Request.hs (hunk 3 above)
Comment (fumieval) [reply src/Network/Wai/SAML2/Request.hs]: Oh sorry, I wasn't reading the comment carefully... I've just updated the code so that it uses `timeFormat` exported from `Network.Wai.SAML2.XML`

**Comment by fumieval** [comment]
I'm afraid I'm not really sure what the issues with documentation and formatting are. Could you reformat the code?

**Comment by mbg** [comment]
No problem and thanks for this contribution! :) I have merged it and pushed [my changes](https://github.com/mbg/wai-saml2/compare/45c45dfc610987a952c29c3e7a320861fd0c7c77..ef756fafbf13e5868ba7e264204bd2aa5c8b3882) on top. You may wish to have a look over those to make sure all is good with you as well.

**Comment by fumieval** [comment]
Thank you for making up the documentation!
